Add gitlab_users datasource

jorcau opened the pull request:
Add new gitlab_users resource that returns a list of users. It takes optional arguments:
- order_by
- sort
- search
- active
- blocked
- identities_extern_uid (must be set with provider)
- identities_provider (must be set with extern_uid)
- created_before
- created_after
- options

Examples:
* Get all users
```
data "gitlab_users" "all_users" {
}
```
* Get users created before the 5th of April 2018 ordered by username
```
data "gitlab_users" "username" {
    order_by = "username"
    created_before = "2018-04-05"
}
```

pdecat commented on gitlab/data_source_gitlab_users.go:
```diff
@@ -11,25 +16,62 @@ func dataSourceGitlabUsers() *schema.Resource {
 		Schema: map[string]*schema.Schema{
-			"order_by": {
-				Type:     schema.TypeString,
+			"options": {
```
Why nest the search options in an `options` map? This could make sense if several `options` blocks could be defined but this isn't the case here.

jorcau replied:
I wanted to avoid changing the parameter "provider" But I agree it makes more sense having all options at the root.

pdecat commented on gitlab/data_source_gitlab_users.go:
```diff
@@ -11,25 +16,62 @@ func dataSourceGitlabUsers() *schema.Resource {
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
-						"id": {
+						"user_id": {
```
Why not keep the `id` name to stay consistent with the gitlab API and the sort option specified at line 28?

jorcau replied:
It was to match the other data source gitlab_user.

pdecat replied:
You mean this one https://github.com/claranet/terraform-provider-gitlab/pull/4/files#diff-e5c640cec25ca3e10eb4afecef78833dR15

As both are new, I'd rather use `id` in both

pdecat commented on gitlab/data_source_gitlab_users.go:
```diff
@@ -83,21 +137,20 @@ func dataSourceGitlabUsers() *schema.Resource {
-
+	log.Printf("\n\n\nListOptions\n%v\n\n\n", listUsersOptions)
 	users, _, err := client.Users.ListUsers(listUsersOptions)
+	log.Printf("\n\n\nListUsers\n%v\n\n\n", users)
```
To remove or put in trace level logging.

pdecat commented on gitlab/data_source_gitlab_users.go:
```diff
@@ -117,14 +170,76 @@ func flattenGitlabUsers(users []*gitlab.User) []interface{} {
+		options += createdBefore
+	}
+	if createdAfter := data["created_after"].(string); createdAfter != "" {
+		// date, err := time.Parse("2006-01-02", createdAfter)
```
WIP?

jorcau replied:
The field created_after is not yet implemented in the client, so I commented the code (same for provider). Should it be better to totally remove it?

pdecat replied:
I see two options:

- fork the client, add it, use the fork here using dep aliasing (until the change is upstreamed) and un-comment the code
- remove it

jorcau replied:
Got it! Thanks!

jorcau replied:
I just made a PR https://github.com/xanzy/go-gitlab/pull/388.

- [x] Update vendor when client PR will be merged.

jorcau replied:
It's already been merged so I also made the vendor update

pdecat commented on gitlab/data_source_gitlab_users.go:
```diff
@@ -117,14 +170,76 @@ func flattenGitlabUsers(users []*gitlab.User) []interface{} {
+
+	data := d[0].(map[string]interface{})
+	listUsersOptions := &gitlab.ListUsersOptions{}
+	options := ""
```
I'd name this `optionsHash` to make it more explicit.

pdecat commented on gitlab/data_source_gitlab_users.go:
```diff
@@ -117,14 +170,76 @@ func flattenGitlabUsers(users []*gitlab.User) []interface{} {
+	}
+	if search := data["search"].(string); search != "" {
+		listUsersOptions.Search = &search
+		options += search
```
What about hashing user input to reduce the risk of not detecting changes?

I'm thinking about clashes between searches like `metrue` with `active = false` and `me` with `active = true`.

Inserting a separator like `,` between concatenated values may also help.

jorcau replied:
I'm not sure what you mean here. I'm actually hashing user input in order so I don't see how there could be a conflict.

pdecat replied:
I mean that, as is, the following user inputs generate the same `metrue` hash unless I'm missing something:

```
{
 search = "metrue"
 active = false
}
```

```
{
 search = "me"
 active = true
}
```

In this example, this happens because active is not added to the hash if false.

Adding a separator is one workaround, but not the only one. Hashing all values should also work.

jorcau replied:
Thanks for the explanations, I added separators.

jorcau commented:
I moved options to root, removed debug lines, renamed:
- user_id to id
- options to optionsHash
- provider to identities_provider (to avoid conflict with terraform reserved keyword provider)
- extern_uid to identities_extern_uid (to be consistent with provider parameter)
